In [2]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn import datasets
import seaborn as sns

In [15]:
from sklearn.svm import SVC
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.metrics import accuracy_score, classification_report

In [3]:
# Memuat dataset Titanic
titanic = sns.load_dataset("titanic")
titanic

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

NB: Sklearn tidak punya desc

In [5]:
titanic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   survived     891 non-null    int64   
 1   pclass       891 non-null    int64   
 2   sex          891 non-null    object  
 3   age          714 non-null    float64 
 4   sibsp        891 non-null    int64   
 5   parch        891 non-null    int64   
 6   fare         891 non-null    float64 
 7   embarked     889 non-null    object  
 8   class        891 non-null    category
 9   who          891 non-null    object  
 10  adult_male   891 non-null    bool    
 11  deck         203 non-null    category
 12  embark_town  889 non-null    object  
 13  alive        891 non-null    object  
 14  alone        891 non-null    bool    
dtypes: bool(2), category(2), float64(2), int64(4), object(5)
memory usage: 80.7+ KB


In [44]:
# Informasi Atribut
print("Nama Kolom:")
print(titanic.columns.tolist(), "\n")

Nama Kolom:
['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare', 'embarked', 'class', 'who', 'adult_male', 'deck', 'embark_town', 'alive', 'alone'] 



# Membuat contoh dataframe dari dataset bawaan

In [10]:
# Memuat dataset kanker payudara
bc = datasets.load_breast_cancer()

# Membuat DataFrame dari fitur-fitur data
dbc = pd.DataFrame(bc.data, columns=bc.feature_names)

# Menambahkan kolom target sebagai kolom baru
dbc['target'] = bc.target

In [11]:
dbc

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0          17.99         10.38          122.80     1001.0          0.11840   
1          20.57         17.77          132.90     1326.0          0.08474   
2          19.69         21.25          130.00     1203.0          0.10960   
3          11.42         20.38           77.58      386.1          0.14250   
4          20.29         14.34          135.10     1297.0          0.10030   
..           ...           ...             ...        ...              ...   
564        21.56         22.39          142.00     1479.0          0.11100   
565        20.13         28.25          131.20     1261.0          0.09780   
566        16.60         28.08          108.30      858.1          0.08455   
567        20.60         29.33          140.10     1265.0          0.11780   
568         7.76         24.54           47.92      181.0          0.05263   

     mean compactness  mean concavity  mean concave points  mean symmetry  \
0             0.27760         0.30010              0.14710         0.2419   
1             0.07864         0.08690              0.07017         0.1812   
2             0.15990         0.19740              0.12790         0.2069   
3             0.28390         0.24140              0.10520         0.2597   
4             0.13280         0.19800              0.10430         0.1809   
..                ...             ...                  ...            ...   
564           0.11590         0.24390              0.13890         0.1726   
565           0.10340         0.14400              0.09791         0.1752   
566           0.10230         0.09251              0.05302         0.1590   
567           0.27700         0.35140              0.15200         0.2397   
568           0.04362         0.00000              0.00000         0.1587   

     mean fractal dimension  ...  worst texture  worst perimeter  worst area  \
0                   0.07871  ...          17.33           184.60      2019.0   
1                   0.05667  ...          23.41           158.80      1956.0   
2                   0.05999  ...          25.53           152.50      1709.0   
3                   0.09744  ...          26.50            98.87       567.7   
4                   0.05883  ...          16.67           152.20      1575.0   
..                      ...  ...            ...              ...         ...   
564                 0.05623  ...          26.40           166.10      2027.0   
565                 0.05533  ...          38.25           155.00      1731.0   
566                 0.05648  ...          34.12           126.70      1124.0   
567                 0.07016  ...          39.42           184.60      1821.0   
568                 0.05884  ...          30.37            59.16       268.6   

     worst smoothness  worst compactness  worst concavity  \
0             0.16220            0.66560           0.7119   
1             0.12380            0.18660           0.2416   
2             0.14440            0.42450           0.4504   
3             0.20980            0.86630           0.6869   
4             0.13740            0.20500           0.4000   
..                ...                ...              ...   
564           0.14100            0.21130           0.4107   
565           0.11660            0.19220           0.3215   
566           0.11390            0.30940           0.3403   
567           0.16500            0.86810           0.9387   
568           0.08996            0.06444           0.0000   

     worst concave points  worst symmetry  worst fractal dimension  target  
0                  0.2654          0.4601                  0.11890       0  
1                  0.1860          0.2750                  0.08902       0  
2                  0.2430          0.3613                  0.08758       0  
3                  0.2575          0.6638                  0.17300       0  
4                  0.1625          0.2364                  0.076

# Data Preparation

In [19]:
# Menghapus baris dengan nilai yang hilang
titanic = titanic.dropna()  # Drop baris dengan NaN di kolom penting
titanic

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
1           1       1  female  38.0      1      0  71.2833        C  First   
3           1       1  female  35.0      1      0  53.1000        S  First   
6           0       1    male  54.0      0      0  51.8625        S  First   
10          1       3  female   4.0      1      1  16.7000        S  Third   
11          1       1  female  58.0      0      0  26.5500        S  First   
..        ...     ...     ...   ...    ...    ...      ...      ...    ...   
871         1       1  female  47.0      1      1  52.5542        S  First   
872         0       1    male  33.0      0      0   5.0000        S  First   
879         1       1  female  56.0      0      1  83.1583        C  First   
887         1       1  female  19.0      0      0  30.0000        S  First   
889         1       1    male  26.0      0      0  30.0000        C  First   

       who  adult_male deck  embark_town alive  alone  
1    woman       False    C    Cherbourg   yes  False  
3    woman       False    C  Southampton   yes  False  
6      man        True    E  Southampton    no   True  
10   child       False    G  Southampton   yes  False  
11   woman       False    C  Southampton   yes   True  
..     ...         ...  ...          ...   ...    ...  
871  woman       False    D  Southampton   yes  False  
872    man        True    B  Southampton    no   True  
879  woman       False    C    Cherbourg   yes  False  
887  woman       False    B  Southampton   yes   True  
889    man        True    C    Cherbourg   yes   True  

[182 rows x 15 columns]

In [22]:
# Menggunakan One-Hot Encoding untuk variabel kategorik
tc = titanic[["sex", "embarked", "class", "embark_town", "who"]]
tn = titanic[["age", "fare", "sibsp", "parch"]]
tn

age     fare  sibsp  parch
1    38.0  71.2833      1      0
3    35.0  53.1000      1      0
6    54.0  51.8625      0      0
10    4.0  16.7000      1      1
11   58.0  26.5500      0      0
..    ...      ...    ...    ...
871  47.0  52.5542      1      1
872  33.0   5.0000      0      0
879  56.0  83.1583      0      1
887  19.0  30.0000      0      0
889  26.0  30.0000      0      0

[182 rows x 4 columns]

# Splitting Data

In [29]:
# Pisahkan fitur dan target
X = tn  # Fitur
y = titanic["survived"]  # Target

In [34]:
# Pisahkan data menjadi data latih dan data uji
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=1, stratify=y)

# Normalisasi data menggunakan StandardScaler
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

print(X_train_scaled)

[[ 6.30742471e-01  5.10111968e-02 -7.92964913e-01 -6.18760096e-01]
 [ 5.45526164e-03  6.24782550e+00 -7.92964913e-01  7.13147229e-01]
 [-8.07418111e-01  5.38524962e-01  9.13925662e-01 -6.18760096e-01]
 [-1.99546381e+00 -5.24133436e-01  2.62081624e+00  7.13147229e-01]
 [ 1.81878817e+00 -1.95368174e-01 -7.92964913e-01  7.13147229e-01]
 [-6.82360669e-01  2.20967867e-01  9.13925662e-01 -6.18760096e-01]
 [ 5.68213750e-01  1.12173966e-01  9.13925662e-01 -6.18760096e-01]
 [ 1.25602968e+00  1.07642913e-01 -7.92964913e-01  7.13147229e-01]
 [ 2.55570146e-01 -6.38590080e-01 -7.92964913e-01 -6.18760096e-01]
 [ 4.43156309e-01  1.94151304e+00 -7.92964913e-01  7.13147229e-01]
 [-2.18304997e+00 -5.24133436e-01  2.62081624e+00  7.13147229e-01]
 [ 2.55570146e-01 -1.08210957e+00 -7.92964913e-01 -6.18760096e-01]
 [ 1.19350096e+00 -6.45743620e-01 -7.92964913e-01 -6.18760096e-01]
 [ 5.45526164e-03  6.34740078e-01  9.13925662e-01  2.04505455e+00]
 [-1.24511916e+00  1.55452884e-01 -7.92964913e-01 -6.18760096e

# Fitting & Evaluation SVM Model

In [38]:
from sklearn.metrics import confusion_matrix, accuracy_score, precision_score, recall_score, f1_score

# Buat objek model SVM dengan kernel linear
model = SVC(kernel='linear')

# Lakukan fitting model dengan data latih
model.fit(X_train_scaled, y_train)

# Lakukan prediksi pada data uji
y_pred = model.predict(X_test_scaled)

# Evaluasi performa model dengan Confusion Matrix dan metrik evaluasi
cm = confusion_matrix(y_test, y_pred)
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred)
recall = recall_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)

print("Parameter Model:", model.get_params(), "\n")
print("Koefisien Model (Jika Linear):", model.coef_ if hasattr(model, "coef_") else "N/A", "\n")
print("Confusion Matrix:")
print(cm, "\n")
print("Accuracy:", accuracy)
print("Precision:", precision)
print("Recall:", recall)
print("F1 Score:", f1)

Parameter Model: {'C': 1.0, 'break_ties': False, 'cache_size': 200, 'class_weight': None, 'coef0': 0.0, 'decision_function_shape': 'ovr', 'degree': 3, 'gamma': 'scale', 'kernel': 'linear', 'max_iter': -1, 'probability': False, 'random_state': None, 'shrinking': True, 'tol': 0.001, 'verbose': False} 

Koefisien Model (Jika Linear): [[-2.12544511e-04  4.80304599e-04  4.08887674e-05 -2.05623295e-04]] 

Confusion Matrix:
[[ 0 18]
 [ 0 37]] 

Accuracy: 0.6727272727272727
Precision: 0.6727272727272727
Recall: 1.0
F1 Score: 0.8043478260869564


model.get_params(): Menampilkan parameter default untuk SVM dengan kernel linear. Parameter ini bisa diatur untuk mengubah perilaku model.

model.coef_: Koefisien dari model SVM dengan kernel linear, koefisien ini memberikan informasi tentang orientasi dan bobot vektor pendukung. Dalam contoh ini,kita memiliki empat koefisien yang dapat membantu menginterpretasikan bagaimana fitur yang berbeda berkontribusi terhadap keputusan klasifikasi.

Confusion Matrix <br>
Baris pertama (kelas 0) menunjukkan bahwa semua 18 prediksi adalah kesalahan (false negatives), tidak ada true positives. <br>
Baris kedua (kelas 1) menunjukkan bahwa semua 37 prediksi adalah benar (true positives), tidak ada false negatives. <br>
Model ini sangat bias ke satu kelas (kelas 1), mungkin karena ketidakseimbangan data atau batasan dalam pemisahan linear.

Precision dan recall memberikan informasi tambahan yang berguna dalam kasus ketidakseimbangan kelas. Model ini memiliki recall tinggi, tetapi precision yang lebih rendah, menunjukkan banyak false positives. <br>
F1 score tinggi menunjukkan keseimbangan yang baik antara precision dan recall.

# Feature Selection & Regularization

In [41]:
from sklearn.feature_selection import SelectKBest, f_classif
from sklearn.model_selection import GridSearchCV

# Buat objek model SVM
model = SVC()

# Gunakan SelectKBest untuk memilih 5 fitur terbaik
selector = SelectKBest(score_func=f_classif, k=4)
X_train_selected = selector.fit_transform(X_train_scaled, y_train)
X_test_selected = selector.transform(X_test_scaled)

# Lakukan Grid Search untuk mencari parameter C terbaik
param_grid = {'C': [0.1, 1, 10]}
grid_search = GridSearchCV(model, param_grid, cv=5)
grid_search.fit(X_train_selected, y_train)

# Hasil terbaik dari Grid Search
best_params = grid_search.best_params_
best_model = grid_search.best_estimator_

# Evaluasi performa model terbaik pada data uji
accuracy = best_model.score(X_test_selected, y_test)

print("Best Parameters:", best_params)
print("Accuracy on Test Data:", accuracy)

Best Parameters: {'C': 0.1}
Accuracy on Test Data: 0.6727272727272727


Hasil Grid Search menunjukkan bahwa regularisasi yang kuat (dengan C=0.1) adalah parameter terbaik untuk model SVM dengan 4 fitur terbaik.

Akurasi pada data uji menunjukkan bahwa model dapat bekerja dengan baik, namun masih ada ruang untuk perbaikan.

# Cross Validation

In [42]:
from sklearn.model_selection import cross_val_score, GridSearchCV

model_cv = SVC(kernel='linear')
# Lakukan K-Fold Cross Validation dengan k=5
scores_cv = cross_val_score(model_cv, X_train_scaled, y_train, cv=5)

# Hitung rata-rata akurasi dari hasil Cross Validation
mean_accuracy_cv = scores_cv.mean()

print("Cross Validation Scores:", scores_cv)
print("Mean Accuracy from Cross Validation:", mean_accuracy_cv)

Cross Validation Scores: [0.65384615 0.69230769 0.68       0.68       0.68      ]
Mean Accuracy from Cross Validation: 0.6772307692307693


Cross Validation Scores: [0.65384615, 0.69230769, 0.68, 0.68, 0.68]:
merupakan akurasi pada setiap lipatan dari validasi silang.
Nilai akurasi yang bervariasi menunjukkan bahwa model mungkin memiliki performa yang berbeda-beda tergantung pada bagian data yang digunakan sebagai validasi.

Mean Accuracy from Cross Validation: 0.6772307692307693: merupakan rata-rata akurasi dari 5 lipatan.
Rata-rata ini memberikan estimasi umum tentang performa model.
Rata-rata akur

# Performance tunning

In [43]:
model_tune = SVC()

# Tentukan parameter yang akan di-tune dan nilai-nilai yang akan dicoba
param_grid = {'C': [0.1, 1, 10], 'kernel': ['linear', 'rbf']}

# Lakukan Grid Search untuk mencari parameter terbaik
grid_search = GridSearchCV(model_tune, param_grid, cv=5)
grid_search.fit(X_train_scaled, y_train)

# Hasil terbaik dari Grid Search
best_params = grid_search.best_params_
best_model = grid_search.best_estimator_

# Evaluasi performa model terbaik pada data uji
accuracy_tune = best_model.score(X_test_scaled, y_test)

print("Best Parameters:", best_params)
print("Best Model:", best_model)
print("Accuracy on Test Data (after tuning):", accuracy_tune)

Best Parameters: {'C': 10, 'kernel': 'linear'}
Best Model: SVC(C=10, kernel='linear')
Accuracy on Test Data (after tuning): 0.6727272727272727


Parameter terbaik untuk SVC (Support Vector Machine) dengan regularisasi yang lemah (C=10) dan kernel linear memberikan kinerja terbaik dalam Grid Search.
Dengan parameter tersebut, model memiliki akurasi sekitar 67.27% pada data uji, yang bisa menjadi indikator kinerja yang baik tetapi mungkin perlu dievaluasi lebih lanjut untuk menilai generalisasi.